### ISSUE TITLE
Can we get OPUS Support anytime soon?

### ISSUE THREAD

**Issue by PlastMan420** [naudio/NAudio#674]

**Comment by markheath** [comment]
If you want opus support in .NET the only managed port I'm aware of is [concentus](https://github.com/lostromb/concentus) which sadly hasn't been updated in a while.
Otherwise the best option is probably to create PInvoke wrappers for a native Opus library

**Comment by prajaybasu** [comment]
FYI, modern Windows 10 installs ship with an Opus decoder MFT (CLSID 63E17C10-2D43-4C42-8FE3-8D8B63E46A6A)

**Comment by rampaa** [comment]
Windows 10 does come with an Opus decoder but when I try to read an Opus file with `StreamMediaFoundationReader` I get the following exception: `System.Runtime.InteropServices.COMException: The byte stream type of the given URL is unsupported. (0xC00D36C4)'`.

Here's code that gives me the aforementioned exception: `IWaveProvider waveProvider = new StreamMediaFoundationReader(new MemoryStream(audio))`.

@markheath do you have any idea why I get the aforementioned exception? And any tips to resolve it?

**Comment by markheath** [comment]
Do you need to use `StreamMediaFoundationReader`? Can you provide the path or a URL instead - that tends to work better. I think the file extension is useful information to know which container format the audio is in.

**Comment by rampaa** [comment]
> Do you need to use StreamMediaFoundationReader?

Yes because I need to check the hash of the audio before playing it so I read it into a byte array first.

Nonetheless, I've tried it with `MediaFoundationReader` as well just to see if it works but unfortunately I still get the `The byte stream type of the given URL is unsupported. (0xC00D36C4)'` exception.

> I think the file extension is useful information to know which container format the audio is in.

The file extension is `.opus`. Here's a sample audio file: https://gofile.io/d/6lkHcL

**Comment by markheath** [comment]
Yes, doesn't seem that `MFCreateSourceReaderFromURL` works with the `.opus` extension. If I had to guess I think maybe using `MFCreateSourceResolverFromURL` and experimenting with flags like `MF_RESOLUTION_CONTENT_DOES_NOT_HAVE_TO_MATCH_EXTENSION_OR_MIME_TYPE` might create an `IMFSource`, and then perhaps passing that into `MFCreateSourceReaderFromMediaSource` would work. Afraid I don't really have the time to work through all that at  the moment though

**Comment by markheath** [comment]
Good news: the upcoming NAudio 3 (currently in development on `main`) adds a new cross-platform package, **`NAudio.SoundFile`**, that supports Opus directly — both decoding and encoding. It's backed by libsndfile, so `SoundFileReader` will decode an `.opus` file as a normal repositionable `WaveStream` (32-bit float), and `SoundFileWriter` can encode Opus as well. Because it goes through libsndfile rather than Media Foundation, it avoids the `0xC00D36C4` "byte stream type is unsupported" error that `MediaFoundationReader`/`StreamMediaFoundationReader` give on `.opus` containers, and it works on Linux and macOS too — not just Windows.

A couple of caveats: Opus is an optional libsndfile build feature, so you'll need a system libsndfile built with Opus support (libsndfile ≥ 1.0.29 + libopus); on Windows that means shipping a suitable libsndfile build alongside your app. NAudio 3 requires .NET 9 or later, so if you're on .NET Framework or .NET 6–8 you'd stay on NAudio 2.x, where the practical options remain the managed Concentus port or a native libopus PInvoke wrapper.

Given that NAudio 3 covers this, I'm planning to close this request as addressed.
